enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,486 @@
+
+While the target installation requires exactly 2 nodes, this will be achieved by proving out the "bootstrap plus 2 nodes" flow in the core installer and then using the Assisted Installer's ability to bootstrap from one of the target machines to remove the requirement for a bootstrap node.
+
+So far, we've discovered topology-sensitive logic in ingress, authentication, CEO, and the cluster-control-plane-machineset-operator. We expect to find others once we introduce the new infrastructure topology.
+
+The delivery of RHEL-HA components will be opaque to the user and be delivered as an [MCO Extension](../rhcos/extensions.md) in the 4.18 and 4.19 timeframes.
+A switch to [MCO Layering](../ocp-coreos-layering/ocp-coreos-layering.md ) will be investigated once it is GA in a shipping version of OpenShift.
+
+Once installed, the configuration of the RHEL-HA components will be done via an in-cluster entity. This entity could be a dedicated in-cluster operator or a function of CEO triggering a script on one of the control-plane nodes. This initialization will require that RedFish details have been collected by the installer.
+Sensible defaults will be chosen where possible, and user customization only where necessary.
+
+This RHEL-HA initialization script will also configure a fencing priority.
+This is usually done based on the sort order of a piece of shared info (such as IP or node name).
+The priority takes the form of a delay, usually in the order of 10s of seconds, and is used to prevent parallel fencing operations during a primary-network outage where each side powers off the other - resulting in a total cluster outage.
+
+RHEL-HA has no real understanding of the resources (IP addresses, file systems, databases, even virtual machines) it manages.
+It relies on resource agents to understand how to check the state of a resource, as well as start and stop them to achieve the desired target state.
+How a given agent uses these actions, and associated states, to model the resource is opaque to the cluster and depends on the needs of the underlying resource.
+
+Resource agents must conform to one of a variety of standards, including systemd, SYS-V, and OCF.
+The latter is the most powerful, adding the concept of promotion, and demotion.
+More information on creating OCF agents can be found in the upstream [developer guide](https://github.com/ClusterLabs/resource-agents/blob/main/doc/dev-guides/ra-dev-guide.asc).
+
+Tools for extracting support information (must-gather tarballs) will be updated to gather relevant logs for triaging issues.
+
+As part of the fencing setup, the cri-o and kubelet services will still be owned by systemd when running under pacemaker. The main difference is that the resource agent will be responsible for signaling systemd to change their active states. The etcd pods are different in this respect since they will be restarted using Podman. It may be possible to start these with the same user account as the original pods.
```

Comment:
as part of the GA process, do you guys also want to check how the fencing could be done via quorum-restore.sh? As far as I understand the current solution, we're mimicking a similar version with podman

That way we can keep etcd under the management of the operator (still)  - this is certainly becoming relevant as we talk about upgrades for GA..

[1] https://docs.google.com/document/d/1oIlHD-gkkskQUNN7n1EWVq5vi3UCOWJHuY3QEzDKsYY/edit?tab=t.0

Reply:
I'm not sure I understand the question (or specifically how it improves upgrades).

Currently, if a node shuts down gracefully, it would exit from the cluster etcd cluster and then rejoin upon its return when pacemaker restarts the pod. If I understand correctly, an upgrade event is 2 graceful restarts (one for each node). I assume that when the rebooted node tries to join the etcd cluster, it takes a moment to resync before it's admitted as a voting peer. 

Based on a quick read of `quorum-restore.sh`, it sounds like it handles the case of non-graceful shutdown. That is to say - we tell a node to bump its version and force the other nodes to rejoin it.

Reply:
My assumption was that you could, potentially, leave etcd under the entire management of CEO when you leverage this script instead of the podman container commands. This should give you an easier path to enabling upgrades.


I don't really get what the difference for etcd in terms of graceful/non-graceful shutdowns would be. it's always member removal and/or force-new-cluster.

Reply:
> I don't really get what the difference for etcd in terms of graceful/non-graceful shutdowns would be

from the prospective of each node taking care of its own etcd instance only, the graceful shutdown allows the agent to leave the member list, and allows the peer to continue without restart a new cluster.

I guess that when there is a central controller (CEO), this difference does not exist

Reply:
:) 

The CEO is very likely not able to tell when a node is shutting down, so I think this invariant is already breaking the assumption. Thanks for the clarification!